README.md
# Welcome to PiDeck
Turn your Pimoroni Pico RBG Keypad into a USB Shortcut Keypad that is configurable via a simple Yaml file.

# Installation & Prerequisites
To get up and running quickly, 
1. Flash your Pico with the latest version of [Circuit Python](https://circuitpython.org/board/raspberry_pi_pico/), from Adafruit
1. Clone this repository and copy all the files on to the pico
2. Edit the `keypad.yml` file to setup your own keyboard shortcuts
3. Restart the Pico and away you go.

# keypad.yml
You can configure the PiDeck using the `keypad.yml` file. It has 16 sections, one for each key.

- Each key as a `name` 0 to 15 (why not 1 to 16? - it's a programming thing!)
- Then it has an `on` and `off` HEX RGB value (without any leading `#` character). The off value will be shown when the key isn't press, the on value when the key is pressed. 
- If the `effect` is set to `pulse` the key will pulse between the off and on values.
- The `command` is the sequence of keys (or shortcut) that will be sent to the computer. On a Mac, you can use the Automator to specify a `service` and then Application to launch, then in the Keyboard panel with Settings you can specify the shortcut key combination to launch that app. Pair that with the key combination in the `command` and you have yourself a productivity Swiss Army knife.
- Finally `button_type` lets you choose whether the key is a single `press` momentry touch button, or if it is a `toggle` button where it stays on until you press it again.
- Optional: the `pulse_timing` will determine how many ticks of the program the `effect` will span.  This must be a positive integer greater than 0.
- Optional: the `repeatable` boolean will allow a button to be pressed and the command to be sent each tick of the program.  Anyone order spam?


``` yaml
- name: 0
   off: FFFFFF
   on: B4F8C8
   effect: none
   command: COMMAND S
   button_type: press
- name: 1
   off: FFFF00
   on: 00FF00
   effect: pulse
   pulse_timing: 1
   command: COMMAND TAB
   button_type: toggle
   repeatable: true
```


CHANGELOG.md
# CHANGELOG

## [1.1] 2022-10-21
### Added
- Optional configuration `pulse_timing` that can be set to control the speed of the `effect`.
- Optional configuration `repeatable` that will allow a button to be pressed and the command to be repeatedly sent.


### Changed
- Altered the color change animations when pressing the buttons.  The button will hold the color on a toggle to be more natural feeling.
- Altered the color animations on toggle buttons to only animate when toggled `on`.
- Added the PRINT_SCREEN key.


code.py
# Adapted from Sandy J Macdonald's gist at https://gist.github.com/sandyjmacdonald/b465377dc11a8c83a8c40d1c9b990a90 to configure all buttons and switch off all lights in loop
# Adapted from Colin Deady's solution on Pimoroni Forum - https://forums.pimoroni.com/t/pico-rgb-keypad-base-as-a-usb-hid-device-solved/16224/7

"""
Purpose: Read in a YAML configuration file, which lists:
- The Key
- The Key On colour and brightness
- The Key Off colour and brightness
- The 'command' to run: its just a text string to send

Maybe also add some pulsing effects too
"""

import time
import board
import busio
import usb_hid

from adafruit_bus_device.i2c_device import I2CDevice
import adafruit_dotstar

from adafruit_hid.keyboard import Keyboard
from adafruit_hid.keyboard_layout_us import KeyboardLayoutUS
from adafruit_hid.keycode import Keycode

from adafruit_hid.consumer_control import ConsumerControl
from adafruit_hid.consumer_control_code import ConsumerControlCode

from digitalio import DigitalInOut, Direction, Pull

import qd_yaml
from key import Key, convert_hex_to_rgb, convert_rgb_to_hex

VERSION = 1.1

cs = DigitalInOut(board.GP17)
cs.direction = Direction.OUTPUT
cs.value = 0
num_pixels = 16
pixels = adafruit_dotstar.DotStar(board.GP18, board.GP19, num_pixels, brightness=0.1, auto_write=True)
i2c = busio.I2C(board.GP5, board.GP4)
device = I2CDevice(i2c, 0x20)
kbd = Keyboard(usb_hid.devices)
layout = KeyboardLayoutUS(kbd)

CONFIG_FILE = 'keypad.yml'

def load_configuration():
    """ Read in the YAML configuration file, and returns the config as a list """
    yml = qd_yaml.YAML()
    with open(CONFIG_FILE) as file:
        config = yml.load(file)

    return config
        

def colourwheel(pos):
    "Change the colour based on the current RGB value"
    if pos < 0 or pos > 255:
        return (0, 0, 0)
    if pos < 85:
        return (255 - pos * 3, pos * 3, 0)
    if pos < 170:
        pos -= 85
        return (0, 255 - pos * 3, pos * 3)
    pos -= 170
    return (pos * 3, 0, 255 - pos * 3)

def read_button_states(x, y):
    "Read the button state"
    
    # set all 16 buttons to state 0
    pressed = [0] * 16
    with device:
        device.write(bytes([0x0]))
        result = bytearray(2)
        device.readinto(result)
        b = result[0] | result[1] << 8
        for i in range(x, y):
            if not (1 << i) & b:
                pressed[i] = 1
            else:
                pressed[i] = 0
    return pressed


# held is set of the button is held down, for debouncing
held = [0] * 16

def set_keycolours(config):
#     """ Set the colours based on the yaml file configuration """

    for key in config:
        key = dict(sum(map(list, map(dict.items, key)), []))
        pixels[int(key["name"])] = convert_hex_to_rgb(key["off"])

def key_on(config, key):
    pixels[int(key["name"])] = convert_hex_to_rgb(key["on"])
    
# load the config and set the key colours
config = load_configuration()

set_keycolours(config=config)

keys = []
item_len = len(config)
# print(f"item_len {item_len}")
for index in range(item_len):
#     print(f"index: {index}")
    key = dict(sum(map(list, map(dict.items, config[index])), []))
    key_no = int(key['name'])
    myKey = Key()
    myKey.command = key['command']
    myKey.name = key['name']
    myKey.on = key['on']
    myKey.off = key['off']
    myKey.effect = key['effect']
    myKey.pulse_timing = key.get('pulse_timing')
    myKey.button_type = key['button_type']
    myKey.repeatable = key.get('repeatable', 'false')
    keys.insert(index, myKey)
# print(f"keys: {keys}")

print(f"PiDeck Version {VERSION}")

while True:
    # check the button press state
    pressed = read_button_states(0, 16)
    
    for key_no in range(16):

        do_effect = True
        if keys[key_no].button_type == "toggle":
            do_effect = keys[key_no].toggle_value
        
        if pressed[key_no] and not held[key_no]:
            # print(f"keys[{key_no}]: {keys[key_no].name} {keys[key_no].button_type} keys[key_no].repeatable {keys[key_no].repeatable}")

            if keys[key_no].button_type == "toggle":
                if keys[key_no].toggle:
                    pixels[key_no] = convert_hex_to_rgb(keys[key_no].off)
                    do_effect = False
                else:
                    pixels[key_no] = convert_hex_to_rgb(keys[key_no].on)
            else:
                pixels[key_no] = convert_hex_to_rgb(keys[key_no].on)
            
            # send the command
            keys[key_no].send(kbd)
            # print(f"key sent {keys[key_no].command}")

            held[key_no] = True
        else:
            if keys[key_no].effect == "none":
                if keys[key_no].button_type == "press":
                    pixels[key_no] = convert_hex_to_rgb(keys[key_no].off)

        # Pulse effect if switched on
        if keys[key_no].effect == "pulse" and do_effect and not pressed[key_no]:
            color = keys[key_no].pulse_tick()
            pixels[key_no] = convert_hex_to_rgb(color)
            
        # Flash effect if switched on
        if keys[key_no].effect == "flash" and do_effect and not pressed[key_no]:
            color = keys[key_no].flash_tick()
            pixels[key_no] = convert_hex_to_rgb(color)
         
    # Released state
    time.sleep(0.15) # Debounce
    # check the button press states again
    pressed = read_button_states(0, 16)
    for key_no in range(16): 
        if not pressed[key_no] or keys[key_no].repeatable:
            held[key_no] = False  # Set held states to off
